gax-grpc/src/main/java/com/google/api/gax/grpc/ClientSettings.java

```diff
@@ -27,129 +27,144 @@
-    this.channelProvider = channelProvider;
-    this.credentialsProvider = Preconditions.checkNotNull(credentialsProvider);
-  }
-
-  /** Gets a channel and an executor for making calls. */
-  public final ChannelAndExecutor getChannelAndExecutor() throws IOException {
-    return ChannelAndExecutor.create(executorProvider, channelProvider);
+    this.transportSettings = transportSettings;
+    this.credentialsProvider = credentialsProvider;
+    this.clock = clock;
   }
 
   public final ExecutorProvider getExecutorProvider() {
     return executorProvider;
   }
 
-  public final ChannelProvider getChannelProvider() {
-    return channelProvider;
+  public final TransportSettings getTransportSettings() {
+    return transportSettings;
   }
 
   public final CredentialsProvider getCredentialsProvider() {
     return credentialsProvider;
   }
 
+  public final ApiClock getClock() {
+    return clock;
+  }
+
   public String toString() {
     return MoreObjects.toStringHelper(this)
         .add("executorProvider", executorProvider)
-        .add("channelProvider", channelProvider)
+        .add("transportSettings", transportSettings)
         .add("credentialsProvider", credentialsProvider)
+        .add("clock", clock)
         .toString();
   }
 
   public abstract static class Builder {
 
     private ExecutorProvider executorProvider;
-    private ChannelProvider channelProvider;
     private CredentialsProvider credentialsProvider;
+    private TransportSettings transportSettings;
+    private ApiClock clock;
 
     /** Create a builder from a ClientSettings object. */
     protected Builder(ClientSettings settings) {
       this.executorProvider = settings.executorProvider;
-      this.channelProvider = settings.channelProvider;
+      this.transportSettings = settings.transportSettings;
       this.credentialsProvider = settings.credentialsProvider;
+      this.clock = settings.clock;
+    }
+
+    protected Builder(ClientContext clientContext) {
+      if (clientContext == null) {
+        this.executorProvider = InstantiatingExecutorProvider.newBuilder().build();
+        this.transportSettings = null;
+        this.credentialsProvider = new NoCredentialsProvider();
+        this.clock = CurrentMillisClock.getDefaultClock();
+      } else {
+        this.executorProvider = FixedExecutorProvider.create(clientContext.getExecutor());
+        this.transportSettings =
+            FixedContextTransportSettings.create(clientContext.getTransportContext());
+        this.credentialsProvider = FixedCredentialsProvider.create(clientContext.getCredentials());
+        this.clock = clientContext.getClock();
+      }
     }
 
-    protected Builder(InstantiatingChannelProvider channelProvider) {
-      this.executorProvider = InstantiatingExecutorProvider.newBuilder().build();
-      this.channelProvider = channelProvider;
-      this.credentialsProvider = new NoCredentialsProvider();
+    protected Builder() {
+      this((ClientContext) null);
```

Comment:
Yes - there are two constructors with a single argument, and `null` has no inherent type, so it has to be disambiguated.